### PR TITLE
 *: Define the schema change at the table level

### PR DESCRIPTION
**Pull request by zimulala**
Add a `cacheItemInfos` in the schema validator to cache the updated table IDs, which is produced when the previous item's version is updated to current item's version.
Checking schema expired has 3 steps:
 1. Whether the current schema of the transaction is expired(the old logic)
 2. Whether the latest schema version in schema validator is expired
 3. Whether the related table IDs in the current transaction is in `cacheItemInfos`

### CODE REVIEW

File: domain/schema_validator.go [hunk 1]
```diff
@@ -22,80 +22,169 @@ import (
...
+	s.mux.Unlock()
+}
 
-	// Delete expired items, leaseGrantTime is server current time, actually.
-	for k, expire := range s.items {
-		if leaseGrantTime.After(expire) {
-			delete(s.items, k)
+func hasRelatedTableID(relatedTableIDs, updateTableIDs []int64) bool {
+	for _, tblID := range updateTableIDs {
+		for _, relatedTblID := range relatedTableIDs {
```
Comment (shenli) [inline domain/schema_validator.go]: Could we use a map here?

Comment (zimulala) [inline domain/schema_validator.go]: I think this is not needed. `relatedTableIDs` generally has a few tables.

File: domain/schema_validator.go [hunk 2]
```diff
@@ -22,80 +22,166 @@ import (
...
+func (s *schemaValidator) IsRelatedTablesChanged(txnTS uint64, currVer int64, tableIDs []int64) (bool, error) {
+	s.mux.RLock()
+	defer s.mux.RUnlock()
+
+	if s.cacheItemInfos == nil {
+		log.Infof("the schema validator stopped before judging")
+		return false, ErrInfoSchemaExpired
+	}
+	if s.isAllExpired(txnTS) {
+		log.Infof("the schema validator's latest schema version %d is exprired", s.latestSchemaVer)
```
Comment (hanfei1991) [inline domain/schema_validator.go]: typo 'expired'

File: domain/schema_validator.go [hunk 3]
```diff
@@ -22,80 +22,166 @@ import (
...
 	// The schemaVer is valid within leaseGrantTime plus lease duration.
-	Update(leaseGrantTime uint64, schemaVer int64)
+	// Add the changed table IDs to the new schema information,
+	// which is produced when the oldSchemaVer is updated to the newSchemaVer.
+	Update(leaseGrantTime uint64, oldSchemaVer, newSchemaVer int64, changedTableIDs []int64)
 	// Check is it valid for a transaction to use schemaVer, at timestamp txnTS.
 	Check(txnTS uint64, schemaVer int64) bool
 	// Latest returns the latest schema version it knows, but not necessary a valid one.
 	Latest() int64
+	// IsRelatedTablesChanged returns the result whether realtedTableIDs is changed from usedVer to the latest schema version,
```
Comment (hanfei1991) [inline domain/schema_validator.go]: typo 'related'

File: domain/schema_validator.go [hunk 4]
```diff
@@ -22,80 +22,166 @@ import (
...
 	s.mux.Lock()
 	defer s.mux.Unlock()
-	s.items = make(map[int64]time.Time)
+	s.vaildItems = make(map[int64]time.Time)
+	s.cacheItemInfos = make(map[int64]*deltaSchemaInfo)
 }
 
-func (s *schemaValidator) Update(leaseGrantTS uint64, schemaVer int64) {
+func (s *schemaValidator) Update(leaseGrantTS uint64, oldVer, currVer int64, changedTableIDs []int64) {
 	s.mux.Lock()
```
Comment (hanfei1991) [inline domain/schema_validator.go]: Is it better to defer Unlock()

File: domain/schema_validator.go [hunk 5]
```diff
@@ -22,80 +22,166 @@ import (
...
 }
 
-func newSchemaValidator(lease time.Duration) SchemaValidator {
+// NewSchemaValidator returns a SchemaValidator structure.
+func NewSchemaValidator(lease time.Duration) SchemaValidator {
 	return &schemaValidator{
-		lease: lease,
-		items: make(map[int64]time.Time),
+		lease:          lease,
+		vaildItems:     make(map[int64]time.Time),
```
Comment (hanfei1991) [inline domain/schema_validator.go]: typo `valid`

File: domain/schema_validator.go [hunk 6]
```diff
@@ -22,80 +22,166 @@ import (
...
+		s.cacheItemInfos[currVer].expire = leaseExpire
+	}
+	s.latestSchemaVer = currVer
+	s.latestSchemaInfo = s.cacheItemInfos[currVer]
+
+	// Delete expired cacheItemInfos, leaseGrantTime is server current time, actually.
+	for k, info := range s.cacheItemInfos {
+		if leaseGrantTime.After(info.expire) {
+			delete(s.vaildItems, k)
+			// We cache some expired schema versions to store recently updated table IDs.
```
Comment (hanfei1991) [inline domain/schema_validator.go]: recently -> recent

File: executor/executor.go [hunk 7]
```diff
@@ -263,14 +263,16 @@ func (e *SelectLockExec) Next() (*Row, error) {
...
+		txnCtx := e.ctx.GetSessionVars().TxnCtx
+		txnCtx.ForUpdate = true
 		txn := e.ctx.Txn()
 		for _, k := range row.RowKeys {
 			lockKey := tablecodec.EncodeRowKeyWithHandle(k.Tbl.Meta().ID, k.Handle)
 			err = txn.LockKeys(lockKey)
 			if err != nil {
 				return nil, errors.Trace(err)
 			}
+			txnCtx.UpdateDeltaForTable(k.Tbl.Meta().ID, 0, 0)
```
Comment (hanfei1991) [inline executor/executor.go]: Add some comments here.

File: executor/executor_test.go [hunk 8]
```diff
@@ -2464,6 +2465,66 @@ func (s *testSuite) TestMiscellaneousBuiltin(c *C) {
 	}
 }
 
+func (s *testSuite) TestSchemaCheckerSQL(c *C) {
```
Comment (hanfei1991) [inline executor/executor_test.go]: It seems not a test about any executor. Can we move this test to `session_test.go` ?

File: domain/schema_validator.go [hunk 9]
```diff
@@ -22,80 +22,166 @@ import (
...
 
-func newSchemaValidator(lease time.Duration) SchemaValidator {
+// NewSchemaValidator returns a SchemaValidator structure.
+func NewSchemaValidator(lease time.Duration) SchemaValidator {
 	return &schemaValidator{
-		lease: lease,
-		items: make(map[int64]time.Time),
+		lease:          lease,
+		vaildItems:     make(map[int64]time.Time),
+		cacheItemInfos: make(map[int64]*deltaSchemaInfo),
```
Comment (hanfei1991) [inline domain/schema_validator.go]: `info` is uncountable.

File: domain/schema_validator.go (hunk 4 above)
Comment (zimulala) [inline domain/schema_validator.go]: Never mind. It's old logic.

File: domain/schema_validator.go (hunk 6 above)
Comment (zimulala) [inline domain/schema_validator.go]: @queenypingcap

File: executor/executor_test.go (hunk 8 above)
Comment (zimulala) [inline executor/executor_test.go]: `select for update` in `executor.go`.

File: domain/schema_validator.go (hunk 4 above)
Comment (winkyao) [inline domain/schema_validator.go]: you can refactor it by the way :)

File: executor/executor_test.go [hunk 10]
```diff
@@ -2464,6 +2465,66 @@ func (s *testSuite) TestMiscellaneousBuiltin(c *C) {
...
+	tk.MustExec(`commit;`)
+	// The schema version is out of date in the first transaction, and the SQL can't be retried.
+	tidb.SchemaChangedWithoutRetry = true
+	tk.MustExec(`begin;`)
+	_, err = se1.Execute(`alter table t add index idx1(c);`)
+	c.Assert(err, IsNil)
+	time.Sleep(lease * 2)
+	tk.MustExec(`insert into t values(3, 3);`)
+	_, err = tk.Exec(`commit;`)
+	c.Assert(terror.ErrorEqual(err, domain.ErrInfoSchemaChanged), IsTrue, Commentf("err %v", err))
```
Comment (shenli) [inline executor/executor_test.go]: Why we can not retry the statement?

Comment (shenli) [inline executor/executor_test.go]: I think we should retry and return column number not match error.

Comment (shenli) [inline executor/executor_test.go]: Ignore it.

File: domain/schema_validator.go (hunk 4 above)
Comment (zimulala) [inline domain/schema_validator.go]: It's only two returns, so I think it's OK.

File: domain/schema_validator.go (hunk 6 above)
Comment (queenypingcap) [inline domain/schema_validator.go]: We should use "recently" because it modifies "updated".

File: domain/schema_validator.go [hunk 11]
```diff
@@ -22,80 +22,166 @@ import (
...
-	mux             sync.RWMutex
-	lease           time.Duration
-	items           map[int64]time.Time
-	latestSchemaVer int64
+	mux        sync.RWMutex
+	lease      time.Duration
+	validItems map[int64]time.Time
+	// cacheItemInfos caches the items' information, and some items may be expired.
+	// It's used to cache the updated table IDs, which is produced when the previous item's version is updated to current item's version.
+	cacheItemInfos   map[int64]*deltaSchemaInfo
```
Comment (winkyao) [inline domain/schema_validator.go]: s/cacheItemInfos/deltaSchemaInfos/g ?

File: executor/executor_test.go [hunk 12]
```diff
@@ -1601,6 +1603,91 @@ func (s *testSuite) TestEmptyEnum(c *C) {
...
+			c.Assert(len(list[2]), Equals, 4)
+			c.Assert(len(list[3]), Equals, 4)
+			c.Assert(len(list[4]), Equals, 12)
+		}
+	}
+}
+
+func (s *testSuite) TestSchemaCheckerSQL(c *C) {
+	defer testleak.AfterTest(c)()
+	store, err := tikv.NewMockTikvStore()
```
Comment (tiancaiamao) [inline executor/executor_test.go]: Is it necessary to `NewMockTikvStore`?

File: session.go [hunk 13]
```diff
@@ -219,10 +220,15 @@ func (s *schemaLeaseChecker) Check(txnTS uint64) error {
 func (s *schemaLeaseChecker) checkOnce(txnTS uint64) error {
 	succ := s.SchemaValidator.Check(txnTS, s.schemaVer)
```
Comment (tiancaiamao) [inline session.go]: After accelerate ddl in RC3, some contract does not hold any more, I think `SchemaValidator` have to be redesigned.

Comment (zimulala) [inline session.go]: You are right, I will fix it in the next PR.

File: domain/schema_validator.go [hunk 14]
```diff
@@ -22,80 +22,166 @@ import (
...
+	latestSchemaVer  int64
+	latestSchemaInfo *deltaSchemaInfo
 }
 
-func newSchemaValidator(lease time.Duration) SchemaValidator {
+// NewSchemaValidator returns a SchemaValidator structure.
```
Comment (zimulala) [inline domain/schema_validator.go]: I use it in `session_test.go`.

**Comment by shenli** [comment]
LGTM

File: domain/schema_validator.go (hunk 11 above)
Comment (zimulala) [inline domain/schema_validator.go]: I will change in next PR.

**Pull request merged**